### PR TITLE
[MRG] add generic support for gzipped and zipfile CSVs

### LINKED ISSUE #1903: support reading (and writing) gzipped manifests
ref #1891, https://github.com/sourmash-bio/sourmash/issues/1902

### LINKED ISSUE #2012: support gzip and/or zipped taxonomy CSVs?
we should support .gz for our input CSVs.

we could also support .zip files that contain a SOURMASH-TAXONOMY.csv file or something. Or even have standard names for genbank, gtdb, and lins taxonomies.

### LINKED ISSUE #2154: zipfile databases could include taxonomy information
I think I mentioned this as part of another issue, but I think it would make taxonomy analyses a bit simpler if we included our taxonomy csv (optionally formatted as an sqldb) within the zipfile, similar to how we currently include the manifest.

obviously, this only works for pre-prepared databases, but we could make a database method to index a user-generated taxonomy csv and add it to a sourmash database.

### LINKED ISSUE #2188: should we support gzipped CSVs for all of sourmash?
From a comment in https://github.com/sourmash-bio/sourmash/pull/2178, where we added support for gzipped taxonomy CSVs - why not support csv.gz I/O throughout?

### PR DESCRIPTION
**Pull request by ctb**
Add gzipped CSV input/output and further regularize CSV I/O throughout the codebase.

In brief,
* MOST commands that write CSVs will now gzip the output if a '.gz' extension is provided. The exceptions are commands like `multigather` and `tax *` that build their own filenames.
* MOST commands that read in CSVs will happily handle .gz input. `lca index` is the exception, see https://github.com/sourmash-bio/sourmash/issues/2198 for why.
* MOST commands will properly detect gzip input based only on content, not filename; the major exception is manifest loading, which is only used internally and does not yet use standard CSV handling. Manifest loading will happily load a csv.gz file based on filename.
* importantly, picklists can now be gzipped CSVs 🎉 

This PR also:
* fixes `sourmash prefetch` and `sourmash gather --save-prefetch-csv` to use standard CSV output;
* fixes `sourmash sig check` to use standard CSV loading (now including gz);
* fixes `sourmash sketch fromfile` to use standard CSV loading (now including gz);
* adds explicit tests for `sourmash_args.FileOutputCSV` ;

## Implementation details

FIRST, this PR provides a new context manager, `sourmash_args.FileInputCSV`, that provides a `csv.DictReader` iterator on top of straight CSVs, gzipped CSVs, and zipfiles that contain a specially named file.

For example,
```python
with FileInputCSV('taxonomy.csv.gz') as r:
   for row in r:
       # ...
```
does what you'd expect. More intriguingly,
```python
with FileInputCSV('taxonomy.zip', default_zip_name='SOURMASH-TAXONOMY.csv') as r:
   for row in r:
       # ...
```
will look for `SOURMASH-TAXONOMY.csv` in the given zipfile and load rows from it.

This permits "nice" behavior such as:
```
with sourmash_args.FileInputCSV(filename,
                               default_zip_name="SOURMASH-TAXONOMY.csv") as r:
```
to support text CSV, gzip CSV, and zipfiles containing SOURMASH-TAXONOMY.csv.

In turn, this enables things like distributing taxonomies within zipfile databases per https://github.com/sourmash-bio/sourmash/issues/2154.

The `FileInputCSV` context handler also supports CSVs with version lines (`# KEY: VERSION`), so can load manifests from zip files (although the manifest loading code does not use `FileInputCSV`, because of its reliance on `ZipStorage`).

This PR _also_ adjusts `FileOutputCSV` to automatically gzip the CSV when `.gz` is present on the end of the output filename. Small change (3 lines), many impacts 😁 

Fixes #2188
Fixes #2012 
Fixes https://github.com/sourmash-bio/sourmash/issues/1903
Addresses #2154 but doesn't fix it

## TODO

- [x] explore passing a file handle into `FileInputCSV` for cases where we already know it's a zipfile - implemented but not yet used anywhere
- [x] write a lot more tests for various error conditions 😆 
- [x] consider support for version lines
- [x] document and test the csv.gz _writing_ code throughout the codebase

### CODE REVIEW

**Comment by ctb** [comment]
@bluegenes thoughts welcome :)

File: src/sourmash/sourmash_args.py [hunk 1]
```diff
@@ -651,10 +654,118 @@ def __init__(self, filename):
     def open(self):
         if self.filename == '-' or self.filename is None:
             return sys.stdout
-        self.fp = open(self.filename, 'w', newline='')
+        if self.filename.endswith('.gz'):
```
Comment (ctb) [inline src/sourmash/sourmash_args.py:657]: note to self: this also has wide impact throughout the codebase and should be discussed and tested!!

**Comment by bluegenes** [comment]
As discussed on slack :) -

re: `SOURMASH-TAXONOMY` — would you consider `GTDB-TAXONOMY` and `NCBI-TAXONOMY` instead, with the default being gtdb? 

OR, somewhere in database info/metadata (which we don’t have yet, but have talked about), add the default for that database? In this case, I'm thinking about database info/metadata as database version (e.g. gtdb-rs207), sourmash signature version, creation date, etc -- and then adding default-taxonomy.

**Comment by ctb** [comment]
> As discussed on slack :) -
> 
> re: `SOURMASH-TAXONOMY` — would you consider `GTDB-TAXONOMY` and `NCBI-TAXONOMY` instead, with the default being gtdb?

might be fine to support `SOURMASH-TAXONOMY` now and then (later) `GTDB-TAXONOMY` and `NCBI-TAXONOMY` via command-line switch.

Biggest pragmatic question is when we would start distributing taxonomies with the databases...

**Comment by ctb** [comment]
hmm - @bluegenes what do you think about me ditching the `SOURMASH-TAXONOMY.csv` for _this_ PR, but keeping the generic support for gzip and zipfiles? We could then explore the Right Way to include taxonomies in another PR, which could focus more on the UX, while being built on top of the foundation in this PR?

tl;dr maybe smaller code changes better :)

**Comment by bluegenes** [comment]
> hmm - @bluegenes what do you think about me ditching the `SOURMASH-TAXONOMY.csv` for _this_ PR, but keeping the generic support for gzip and zipfiles? We could then explore the Right Way to include taxonomies in another PR, which could focus more on the UX, while being built on top of the foundation in this PR?
> 
> 
> 
> tl;dr maybe smaller code changes better :)

That'd be ok!

**Comment by ctb** [comment]
Ready for review & merge @sourmash-bio/devs!

File: src/sourmash/sourmash_args.py [hunk 2]
```diff
@@ -651,10 +654,122 @@ def __init__(self, filename):
...
         return self.fp
 
 
+class _DictReader_with_version:
+    """A version of csv.DictReader that allows a comment line with a version,
+    e.g.
+
+    # SOURMASH-MANIFEST-VERSION: 1.0
+
+    The version is stored as a 2-tuple in the 'version_info' attribute.
```
Comment (bluegenes) [inline src/sourmash/sourmash_args.py:670]: yessssssssss

File: src/sourmash/sourmash_args.py [hunk 3]
```diff
@@ -651,10 +654,122 @@ def __init__(self, filename):
...
+            try:
+                zi = zipfile_obj.getinfo(default_csv_name)
+                with zipfile_obj.open(zi) as fp:
+                    textfp = TextIOWrapper(fp,
+                                           encoding=encoding,
+                                           newline="")
+                    r = _DictReader_with_version(textfp, delimiter=delimiter)
+                    yield r
+            except (zipfile.BadZipFile, KeyError):
+                pass # uh oh, we were given a zipfile_obj and it FAILED.
```
Comment (bluegenes) [inline src/sourmash/sourmash_args.py:734]: any fake zip that can cover this line?

File: tests/test_cmd_signature.py [hunk 4]
```diff
@@ -89,6 +95,36 @@ def test_sig_merge_1_fromfile_picklist(runtmp):
...
+    # test with --from-file and gzipped picklist
+    c = runtmp
+
+    # merge of 47 & 63 should be union of mins
+    sig47 = utils.get_test_data('47.fa.sig')
+    sig63 = utils.get_test_data('63.fa.sig')
+    sig47and63 = utils.get_test_data('47+63.fa.sig')
+
+    from_file = _write_file(runtmp, 'list.txt', [sig47, sig63])
+    picklist = _write_file(runtmp, 'pl.csv',
```
Comment (bluegenes) [inline tests/test_cmd_signature.py:108]: ah, so name doesn't matter for output bc you've specified `gz=True`-- writing `pl.csv` but it's gzipped.

Not an issue, just worth mentioning.

Comment (ctb) [reply tests/test_cmd_signature.py:108]: right, this is just a test function. It matters if you use the built-in `FileOutputCSV`.

File: src/sourmash/sourmash_args.py (hunk 3 above)
Comment (ctb) [reply src/sourmash/sourmash_args.py:734]: fixed in c390db9d - thanks!

**Review by bluegenes — APPROVED** [review]
lgtm
